Automattic/newspack-ads
# feat(placements): customize control

Implements the `Placement_Customize_Control` for global placement editing and live refresh through the customizer.

Closes #254 

<img width="1689" alt="image" src="https://user-images.githubusercontent.com/820752/163199857-3a86b3b4-32af-4cc3-8160-aa0f06594ad7.png">

### How to test

1. Check out this branch and make sure you have ad units available
2. Visit your homepage and access **Customize**
3. If you already have global placements with units you'll see them with an edit icon as on the image above
4. Confirm you see the new **Ads Placements** menu item:

<img width="350" alt="image" src="https://user-images.githubusercontent.com/820752/163200528-296883fc-6675-4ecd-8288-b31f459dc945.png">

5. Click on the menu and confirm all the placements are listed:

<img width="342" alt="image" src="https://user-images.githubusercontent.com/820752/163200688-4fedf6c4-31b2-40eb-925e-f3fa56ca3b93.png">

6. Click on the "Global: Above Header" item and select an ad unit for it:
   1. If an ad unit hasn't yet been selected for this placement, a full refresh should take place so the ad unit preview is rendered
   2. Subsequent ad unit changes for the placement should use the [selective refresh method](https://developer.wordpress.org/reference/classes/wp_customize_selective_refresh/add_partial/) and not perform a full refresh
7. Save and confirm the changes are applied to the placement
8. Visit the Customize again, click on the edit icon button from the Above Header placement and confirm you are redirected to its form on the sidebar.

#### Multiple providers

1. Enable Broadstreet with a valid account
2. Edit a placement through the customizer and select Broadstreet as the provider
3. Confirm the preview renders the Broadstreet unit with proper dimensions
4. Save and confirm the change is applied

#### Header bidding

1. Make sure you have header bidding enabled (`define( 'NEWSPACK_ADS_EXPERIMENTAL_BIDDERS', true );`) with debug mode on
5. Edit a placement with GAM provider and add random bidders' IDs
6. Save and confirm the bidders' IDs are applied by checking Prebid console logs
